### PR TITLE
Issue-380 Adding support for '.' in the schema names

### LINKED ISSUE #380: Gen-Design Doc fails to handle type references with periods.
When a field name contains a '.' and is referenced in a type field, the `gen-design-doc` command throws an error.
```
gen-design-doc: general_failure

An unrecognized error occurred:/workspace/AaC/python/src/aac/plugins/gen_design_doc/gen_design_doc_impl.py line 48 message: 'dict object' has no attribute 'SubSchema'
```

This can be replicated with this simple yaml example:
```yaml
schema:
  name: Schema1
  fields:
    - name: data
      type: SubSchema.Schema1
---
schema:
  name: SubSchema.Schema1
  fields:
    - name: data
      type: string
```

## AC:
- [ ] Gen design doc can generate design documents with definitions containing '.' in their names.

### PR DESCRIPTION
**Pull request by JustAChucklin**
This is in relation to this issue ->  this should close #380 

Tests were passing locally, will make changes based on the pipeline's messages if the changes are failing

This is a tentative PR, as I am not sure if the changes that @cvchaparro , and I made would be able to accommodate for certain use-cases/edge-cases that may come up in certain scenarios. 

Let me know if there are any changes that are needing to be made to accommodate those scenarios or if you see something that could accommodate better, and could be merged in to this PR

Thanks in advance!

### CODE REVIEW

File: python/tests/plugins/test_gen_design_doc_impl.py
Comment (cvchaparro) [inline python/tests/plugins/test_gen_design_doc_impl.py:18]: We need to remove the extra spaces here. You may have a Neovim setting that allows removing space at the end of lines - if that's available, please set that (at least for this project). The linter should catch this, too, if I'm not mistaken.

File: python/tests/plugins/test_gen_design_doc_impl.py
Comment (cvchaparro) [inline python/tests/plugins/test_gen_design_doc_impl.py:15]: If you can, it would be good to update this test to follow the pattern we set up in the other test (i.e. instead of manually writing the file with the `with open(...)` part, update the test to use `with temporary_test_file(...)` etc.

File: python/tests/plugins/test_gen_design_doc_impl.py
Comment (cvchaparro) [inline python/tests/plugins/test_gen_design_doc_impl.py:184]: I forgot to mention this while we were working on it but, while we did this to get things working, it would probably be better to use our `create_schema_definition` helper function to create these schemas programmatically so we don't have to store several different constants far from where they're used.

File: python/src/aac/plugins/gen_design_doc/gen_design_doc_impl.py
Comment (cvchaparro) [inline python/src/aac/plugins/gen_design_doc/gen_design_doc_impl.py:104]:
```suggestion
def _get_interfaces_with_required_fields(interfaces: list[dict]) -> list[dict]:

    def add_required_fields_to_interface(interface):
        definition = Definition(structure=interface)
        return interface | {"required_fields": get_required_fields(definition)}

    return [add_required_fields_to_interface(interface) for interface in interfaces]
```

**Review by cvchaparro — CHANGES_REQUESTED** [review]

File: python/tests/plugins/test_gen_design_doc_impl.py
Comment (Coffee2Bits) [reply python/tests/plugins/test_gen_design_doc_impl.py:18]: I believe that you're correct; flake8 should catch this.

Comment (JustAChucklin) [reply python/tests/plugins/test_gen_design_doc_impl.py]: Just pushed some changes through let me know if there is anything else

File: python/tests/plugins/test_gen_design_doc_impl.py
Comment (JustAChucklin) [reply python/tests/plugins/test_gen_design_doc_impl.py]: Updated the first test, please let me know if there is anything else required. Thanks Cameron for that final nudge on getting that portion through :)

File: python/tests/plugins/test_gen_design_doc_impl.py
Comment (JustAChucklin) [reply python/tests/plugins/test_gen_design_doc_impl.py:178]: I am a little apprehensive on that currently. I do think that would be beneficial. But I am not wanting to break anything and causing more delay in getting this added. If it is required I can definitely look into getting those constants removed and exploring the `create_schema_definition` option.

What are your thoughts?

File: python/tests/plugins/test_gen_design_doc_impl.py
Comment (cvchaparro) [inline python/tests/plugins/test_gen_design_doc_impl.py:84]:
```suggestion

```

File: python/tests/plugins/test_gen_design_doc_impl.py
Comment (cvchaparro) [reply python/tests/plugins/test_gen_design_doc_impl.py:178]: That's fine, we can hold off on switching. Instead, please create a story with the label "chore" that addresses updating those test models.

File: python/tests/plugins/test_gen_design_doc_impl.py
Comment (cvchaparro) [inline python/tests/plugins/test_gen_design_doc_impl.py:40]: Please also update this on lines 18-20 as I can't leave it as a suggestion.

```suggestion
            test_design_doc_file_name, *_ = [f for f in files if f != os.path.basename(test_yaml.name)]
```

**Review by cvchaparro — CHANGES_REQUESTED** [review]

File: python/tests/plugins/test_gen_design_doc_impl.py
Comment (JustAChucklin) [reply python/tests/plugins/test_gen_design_doc_impl.py]: Please check the latest commit. I made the above change to the first and second test

File: python/tests/plugins/test_gen_design_doc_impl.py
Comment (JustAChucklin) [reply python/tests/plugins/test_gen_design_doc_impl.py:170]: created issue #421

**Comment by JustAChucklin** [comment]
Just pushed a fix for the linter errors

**Review by cvchaparro — CHANGES_REQUESTED** [review]
Just a couple more clean up suggestions and a comment regarding the `Definition` change we made.

File: python/src/aac/plugins/gen_design_doc/gen_design_doc_impl.py
Comment (Coffee2Bits) [inline python/src/aac/plugins/gen_design_doc/gen_design_doc_impl.py:97]: I'm not able to highlight-comment the function `_get_and_prepare_definitions_by_type`, but this solution would resolve the need to have default values in the `Definition` class and the function `_get_interfaces_with_required_fields`.

```python
def _get_and_prepare_definitions_by_type(parsed_definitions: list[Definition], aac_type: str) -> list[dict]:
    def get_definition_structure_with_required_fields(interface_definition: Definition):
        return interface_definition.structure | {"required_fields": get_required_fields(interface_definition)}

    filtered_definitions = get_definitions_by_root_key(aac_type, parsed_definitions)
    definition_template_properties = []
    if aac_type == "schema":
        definition_template_properties = [get_definition_structure_with_required_fields(definition) for definition in filtered_definitions]
    else:
        definition_template_properties = [definition.structure for definition in filtered_definitions]

    return definition_template_properties
```

Comment (JustAChucklin) [reply python/src/aac/plugins/gen_design_doc/gen_design_doc_impl.py:97]: @cvchaparro , do you want me to go ahead and implement this? What are you thoughts

**Comment by JustAChucklin** [comment]
Pushed that change through @Coffee2Bits , @cvchaparro , and I reverted that first method from the branch. Will I also need to revert back that `datum.md.jinja2` change, or is that still ok to leave?

File: python/src/aac/plugins/gen_design_doc/gen_design_doc_impl.py
Comment (JustAChucklin) [reply python/src/aac/plugins/gen_design_doc/gen_design_doc_impl.py]: Pushed this change through (See recent comment)

File: python/src/aac/lang/definitions/definition.py
Comment (JustAChucklin) [reply python/src/aac/lang/definitions/definition.py]: I think this is resolved now in the latest push. (see last comment)

**Comment by cvchaparro** [comment]
> Pushed that change through @Coffee2Bits , @cvchaparro , and I reverted that first method from the branch. Will I also need to revert back that `datum.md.jinja2` change, or is that still ok to leave?
> 
> Tests were passing locally before I made my last push.
> 

The datum.md.jinja2 change should remain the same.

**Comment by JustAChucklin** [comment]
> > Pushed that change through @Coffee2Bits , @cvchaparro , and I reverted that first method from the branch. Will I also need to revert back that `datum.md.jinja2` change, or is that still ok to leave?
> > Tests were passing locally before I made my last push.
> 
> The datum.md.jinja2 change should remain the same.

Awesome, I think that we are good to go then! Unless there is something else that I missed.

File: python/tests/plugins/test_gen_design_doc_impl.py
Comment (cvchaparro) [inline python/tests/plugins/test_gen_design_doc_impl.py:36]: Since we changed how we handle `required_fields` it would be good to make sure `"required"` doesn't show up in the markdown file contents (since `TEST_MODEL_2` doesn't have any required fields).

```suggestion
                markdown = markdown_file.read()
                self.assertIn("SubSchema.Schema1 data", markdown)
                self.assertNotIn("required", markdown)
```

File: python/src/aac/plugins/gen_design_doc/gen_design_doc_impl.py
Comment (cvchaparro) [inline python/src/aac/plugins/gen_design_doc/gen_design_doc_impl.py:98]: It's cleaner to keep a space around inner functions.

```suggestion

    def get_definition_structure_with_required_fields(interface_definition: Definition):
```

**Review by cvchaparro — CHANGES_REQUESTED** [review]
Looks like you're close but those changes were not complete since there would be dead code left. Also, there are a couple improvements we can make for completeness/cleanliness.

**Comment by JustAChucklin** [comment]
> Looks like you're close but those changes were not complete since there would be dead code left. Also, there are a couple improvements we can make for completeness/cleanliness.

Just pushed the suggestions through that you made in the quoted thread.

**Review by cvchaparro — APPROVED** [review]
Nice job. 😄

**Pull request merged**
